src/main/java/robot/subsystems/turret/Turret.java

```diff
@@ -0,0 +1,149 @@
+package robot.subsystems.turret;
+
+import com.ctre.phoenix.motorcontrol.ControlMode;
+import com.ctre.phoenix.motorcontrol.FeedbackDevice;
+import com.ctre.phoenix.motorcontrol.LimitSwitchNormal;
+import com.ctre.phoenix.motorcontrol.LimitSwitchSource;
+import com.ctre.phoenix.motorcontrol.can.TalonSRX;
+import edu.wpi.first.wpilibj.command.Subsystem;
+import robot.subsystems.turret.commands.TurnTurret;
+
+import static robot.Constants.Turret.*;
+import static robot.Ports.Turret.MASTER;
+
+/**
+ * @author Adam & Barel
+ * @version 1.0
+ * <p>
+ * this class includes methods for the Turret.
+ * {@using TalonSRX}
+ * {@using Relative Encoder}
+ * {@using Hall Effect}
+ */
+public class Turret extends Subsystem {
+    private TalonSRX master = new TalonSRX(MASTER);
+    private double angle;
+    private Direction direction;
+
+    /**
+     * configures the encoder and PID constants.
+     */
+    public Turret() {
+        direction = Direction.RIGHT;
+        master.configSelectedFeedbackSensor(FeedbackDevice.CTRE_MagEncoder_Relative, 0, TIMEOUT_MS);
+        master.config_kP(TALON_PID_SLOT, KP, TIMEOUT_MS);
+        master.config_kI(TALON_PID_SLOT, KI, TIMEOUT_MS);
+        master.config_kD(TALON_PID_SLOT, KD, TIMEOUT_MS);
+        master.config_kF(TALON_PID_SLOT, KF, TIMEOUT_MS);
+        master.setInverted(IS_MASTER_INVERTED);
+        master.configPeakCurrentLimit(MAX_CURRENT);
+        master.configForwardLimitSwitchSource(LimitSwitchSource.FeedbackConnector, LimitSwitchNormal.NormallyOpen);
+        reset();
+        master.setSelectedSensorPosition(0);
+    }
+
+    @Override
+    public void initDefaultCommand() {
+        setDefaultCommand(new TurnTurret(45));
+    }
+
+    /**
+     * set the direction of the turret,
+     * either left or right.
+     *
+     * @param direction the direction that the turret will move.
+     */
+    public void setDirection(Direction direction) {
+        this.direction = direction;
+    }
+
+    /**
+     * get the current angle from the controller
+     *
+     * @return the angle of the turret
+     */
+    public double getAngle() {
+        return convertTicksToDegrees(master.getSelectedSensorPosition());
+    }
+
+    /**
+     * change the angle to the desired angle,
+     * if you would like to use the same Direction.
+     * the value can be between -180 to 180 degrees.
+     *
+     * @param targetAngle the desired angle.
+     */
+    public void setTargetAngle(double targetAngle) {
+        if (Double.compare(targetAngle, angle) == 0) return;
+        while (targetAngle > 180)
+            targetAngle -= 360;
+        while (targetAngle < -180) {
+            targetAngle += 360;

```

Comment:
Do not use `while` loops in FRC code. You risk the code not using in the 20ms it should (the roboRIO loop runs in 50Hz).